flocker/control/_model.py

```diff
@@ -414,6 +425,26 @@ def key_match_invariant(pmap):
 _keys_match_dataset_id = _keys_match("dataset_id")
 
 
+def _turn_iterable_to_mapping_from_attribute(attribute, obj):
+    """
+    Given ``obj`` which is either something that is an instance of ``Mapping``
+    or an iterable, return something that implements ``Mapping`` from
+    ``getattr(o, attribute)`` to the value of the items in the iterable. This
+    funky signature is required to be a valid field factory function.
+
+    :param attribute: The name of the attribute that should be the key in the
+        resulting Mapping.
+    :param obj: Either something that is a Mapping (which is immediately
+        returned), or an iterable to be converted into a Mapping.
+
+    :returns: A Mapping of the items in ``obj`` from
+        ``getattr(o, attribute)`` to the items in obj (``o`` in ``obj``).
+    """
+    if isinstance(obj, Mapping):
+        return obj
+    return {getattr(a, attribute): a for a in obj}
```

Comment:
Hrm... This factory intentionally enables you to pass an iterable of things and turns it into a mapping from an attribute of the objects to the actual objects.

In the use case here, that enables us to hand an iterable of applications, and this function turns it into a mapping from application.name to applications.

If you hand it in an iterable of object that don't have a 'name' attribute... then you should get an attribute error describing that the name attribute is missing... So the AttributeError is definitely the thing that should be raised, and I don't have a mechanism to add context to the exception (... in an iterable handed to the factory of a PClass field).

Expanded the docstring to leave a breadcrumb for people who might encounter this Exception.